swap/src/network/request_response.rs

```diff
@@ -171,3 +193,92 @@ where
         Ok(())
     }
 }
+
+#[derive(Clone, Copy, Debug, Default)]
+pub struct OneShotCodec<P> {
+    phantom: PhantomData<P>,
+}
+
+#[async_trait]
+impl<P> RequestResponseCodec for OneShotCodec<P>
```

Comment:
❓We did the same in `comit-rs` - I am not sure why this is needed though. Is there a short explanation why we have to implement `RequestResponseCodec` for the `OneShotCodec`?

(Is this needed because under the hood the oneshot is still a request-response protocol?)

Reply:
> (Is this needed because under the hood the oneshot is still a request-response protocol?)

Yes, it is using `libp2p-request-response`.
See the `rr` field within `Behaviour` in the next file in this diff.

Reply:
You either have to implement all libp2p abstraction layers yourself, that is:

`ProtocolInfo`, `InboundUpgrade`, `OutboundUpgrade`, `ProtocolsHandler` and `NetworkBehaviour`.

Or you use a more high-level abstraction like `RequestResponse` that takes care of this.

Regardless of which one you are using, you will need to specify somehow, how a message gets encoded and decoded into bytes. Usually you'd do that in either `InboundUpgrade`/`OutboundUpgrade` or in the `ProtocolsHandler` but you don't get in touch with these abstractions when you use `RequestResponse` hence you will need to specify that in a different way: the `RequestResponseCodec` abstraction of `libp2p-request-response`.

I can recommending spending some time trying to implement your own protocol where you touch all the layers (or carefully read the existing code). It will make a lot more sense once you know what is going on behind the scenes.

Reply:
We need a codec for `RequestResponse`.
Yes, the "one shot" is a request (the one shot) and a response (empty, act as ACK) behaviour.